### DISCUSSION TITLE
JSBSim simulation with HOTAS in Unity3D

### DISCUSSION THREAD

**Discussion by pd3a** [JSBSim-Team/jsbsim#879] [Support requests]
I'm developing a flight simulator in Unity3D engine. I want to use JSBSim library to simulate the flight physics. I wrote the wrapper for JSBSim which takes HOTAS inputs through Unity and sends output data (position, velocities, orientation, angular velocities) back to Unity.
The wrapper is working i.e i'm able to get input to jsbsim and output back to unity. In doing so i'm facing the following issues:
1) I'm setting the hotas axis values to 'fcs/"aerosurface"-cmd-norm' but still getting wrong values of orientation the roll and pitch values are in order of e-09 despite giving full input in HOTAS. Once the simulation starts, even after giving full pitch up the altitude keeps dropping.
2) There has been a delay in input commands and outputs received from JSBsim.
I'm aware of existence of copy_to_jsbsim and copy_from_jsbsim but i do not know how to access them in C# for Unity. Help me here please.

**Comment by seanmcleod** [comment]
Okay, so if you're not writing this in C++ and using `copy_to_jsbsim()` and `copy_from_jsbsim()` then how are you interfacing with JSBSim from your C# code? What is the overall structure of your setup?

Are you running JSBSim as a separate process then, as opposed to linking it into your app? And then are you communicating in terms of getting inceptor inputs into JSBSim and state output from JSBSim via a network connection writing and reading properties?

Or have you built JSBSim into a DLL and then you're using something like PInvoke from C# to communicate with JSBSim via your wrapper?

Now I wouldn't set your inceptor's axis scale to the `aerosurface_scale` range. For example, take a look at a typical setup from the 737 model in the JSBSim repo. Set your inceptor's pitch axis scale to `[-1, +1]` and then feed that into the  `fcs/pitch-trim-cmd-norm` property.

https://github.com/JSBSim-Team/jsbsim/blob/10ac2036d8b5ac6d840a105f41062c603e952e6a/aircraft/737/737.xml#L189-L209

**Reply by DreamerSurpass** [reply]
Hello, I also have the same question. I also want to use Unity3D to visualize and send the control input to JSBsim, then JSBSim step forward and refresh the state and Scene in Unity3D.

So is it possible to interface with JSBSim from C# code? In my opinion, I may pack the JSBSim and use dll in C#(but I'm not very familiar with C++) , or run the JSBSim as a seperate process and use the network communication to exchange param.

Which way is more convenient? And how can I interface with JSBSim from C#? Is that possible to use some API (like in python or C++)to set property and get property in C# directly?

Thanks a lot.

**Comment by pd3a** [comment]
Yes, I'm running JSBSim as a separate process and communicating via network connection for reading and writing properties.
But I'm giving inceptor's pitch axis input to fcs/pitch-cmd-norm as of now.

**Comment by seanmcleod** [comment]
Currently there is no C# wrapper along the lines of the Python wrapper for the core JSBSim classes. Since they're C++ classes as opposed to plain C-functions it means you can't simply use P/Invoke to call them from C#, someone would need to write a proper wrapper, probably easiest as a C++/CLI wrapper.

The other option as mentioned by @pd3a is to use a network connection.

**Reply by DreamerSurpass** [reply]
Got it. Thanks for your prompt reply.

**Reply by DreamerSurpass** [reply]
Excuse me I have a new thought. 
Firstly I create JSBSim.lib and link it in my own cpp project. In this project I can write some functions to run JSBSim, and then use `__declspec(dllexport)` to create a dll file. Then link this dll file in C# and use corresponding functions.

Is it feasible?

**Comment by seanmcleod** [comment]
Yes you could add some of your own wrapper functions in C/C++ that are then exported from JSBSim.dll when you build it as a DLL.

Now if you make your wrapper functions plain C functions then it makes it trivial to use P/Invoke from C#. But if you want to expose a C++ class based wrapper then you would need to go with the C++/CLI wrapper approach that I mentioned above.

**Reply by DreamerSurpass** [reply]
Get it. Because I find C++/CLI wrapper less information and  feel it might be more difficult for me. 

And I have tried to wrap test functions in C++ and build a dll. In C#/Unity, some functions works fine, but if I call run() or loadScript() (this might due to String) functions, the program will crash.
![image](https://user-images.githubusercontent.com/93421761/231771588-c43f8e32-b6d5-4306-aebd-a1e6fb1d4cff.png)
![image](https://user-images.githubusercontent.com/93421761/231772994-453e876a-ef4e-4835-b8a4-9c425e243ca9.png)

Do you have any ideas? Is it because JSBSim is running incorrectly?

**Comment by seanmcleod** [comment]
Impossible to tell without seeing your implementation of `createJSB()` and `run()`.

P.S
Rather than posting screen shots of your source code use a triple back-tick ``` followed by C++ and then cut and paste your source code text and then close with another set of triple back-ticks.

Example (use the edit option to see exactly what I typed below):

```c++
void run(FGFDMExec* fdmExec)
{
   fdmExec->run();
}
```

**Reply by DreamerSurpass** [reply]
Alright, thanks for advice. These functions are the simplest implementation to test like this.

```c++
#include "JSBSimWrapper.h"

JSBSimWrapper::JSBSimWrapper()
{
	this->fdm = new FGFDMExec();
}

JSBSimWrapper:: ~JSBSimWrapper()
{

}

void JSBSimWrapper::setdt(float dt)
{
	this->fdm->Setdt(dt);
}

bool JSBSimWrapper::run()
{
	return this->fdm->Run();
}

void JSBSimWrapper::setDebugLevel(int level)
{
	this->fdm->SetDebugLevel(level);
}
```

```c++
#include"JSBSimWrapper.h"

extern "C" {
	_declspec(dllexport) JSBSimWrapper* createJSB();
	_declspec(dllexport) bool run(JSBSimWrapper* jsb);
	_declspec(dllexport) void setDebugLevel(JSBSimWrapper* jsb, int level);
	_declspec(dllexport) void setdt(JSBSimWrapper* jsb, float dt);


	JSBSimWrapper* createJSB()
	{
		return new JSBSimWrapper();
	}

	bool run(JSBSimWrapper* jsb)
	{
		return jsb->run();
	}

	bool setAircraftPath(JSBSimWrapper* jsb, string path)
	{
		return jsb->setAircraftPath(path);
	}

	void setDebugLevel(JSBSimWrapper* jsb, int level)
	{
		jsb->setDebugLevel(level);
	}

	void setdt(JSBSimWrapper* jsb, float dt)
	{
		jsb->setdt(dt);
	}
}
```

**Comment by seanmcleod** [comment]
Not clear in your examples if an aircraft has been sucessfully loaded before `jsb->run()` is called. Use the debugger and see what happens when you step into `this->fdm->Run()`.

**Reply by DreamerSurpass** [reply]
> Not clear in your examples if an aircraft has been sucessfully loaded before `jsb->run()` is called. 

> Yes you could add some of your own wrapper functions in C/C++ that are then exported from JSBSim.dll when you build it as a DLL.

Yes, before, I just wanted to test whether this method works, because I think no aircraft, sript, etc loaded before run will return false or report an error, but the program suddenly crashed. I'll test later. Again, thanks for your help.

**Comment by paulio** [comment]
FYI You can run Python in Unity via Python.NET. It takes a bit of doing, but it does work. Although I'd still look at using a comms pattern. The Unreal JSBSim sample embeds it as a library, which as mentioned is possible if you expose the DLLImport stuff correctly.
